modules/DOMStateStorage.js

```diff
@@ -8,18 +8,33 @@ function createKey(key) {
   return KeyPrefix + key
 }
 
+function isHistoryKey(key) {
+  return key.indexOf(KeyPrefix) === 0
+}
+
 export function saveState(key, state) {
   try {
     window.sessionStorage.setItem(createKey(key), JSON.stringify(state))
   } catch (error) {
-    if (error.name === QuotaExceededError || window.sessionStorage.length === 0) {
-      // Probably in Safari "private mode" where sessionStorage quota is 0. #42
-      warning(
-        false,
-        '[history] Unable to save state; sessionStorage is not available in Safari private mode'
-      )
-
-      return
+    if (error.name === QuotaExceededError) {
+      if (window.sessionStorage.length === 0) {
+        // Probably in Safari "private mode" where sessionStorage quota is 0. #42
+        warning(
+          false,
+          '[history] Unable to save state; sessionStorage is not available in Safari private mode'
+        )
+
+        return
+      }
+
+      // Reached quota, remove first (oldest) item in storage
+      for (const storedKey in window.sessionStorage) {
```

Comment:
It seems like `sessionStorage.key` isn't reliable either, at least not in Chrome.

<img width="261" alt="screen shot 2015-10-15 at 6 08 40 pm" src="https://cloud.githubusercontent.com/assets/92839/10531529/e0aa24de-7367-11e5-9ef8-110962de8f6d.png">